qiskit/optimization/problems/optimization_problem.py

```diff
@@ -342,227 +345,464 @@ def set_problem_type(self, _type):
         qiskit.optimization.problem_type.QCP
         qiskit.optimization.problem_type.MIQCP
         """
-        self.my_problem_type = _type
-
-    def solve(self):
-        """Solves the problem.
+        self._problem_type = problem_type
 
-        Note
-          The solve method returning normally does not necessarily mean
-          that an optimal or feasible solution has been found.  Use
-          OptimizationProblem.solution.get_status() to query the status of the current
-          solution.
+    def solve(self) -> None:
+        """Prints out a message to ask users to use `OptimizationAlgorithm`.
+        Users need to apply one of `OptimiztionAlgorithm`s instead of this method.
         """
-        # TODO: Implement me
-        pass
+        logger.warning('`OptimizationProblem.solve` is intentionally empty.'
+                       'You can solve it by applying `OptimizationAlgorithm.solve`.')
 
-    def set_problem_name(self, name):
-        """Set the problem name."""
+    def set_problem_name(self, name: str):
+        """Sets the problem name"""
         self._name = name
 
     def get_problem_name(self):
-        """Get the problem name."""
+        """Returns the problem name"""
         return self._name
 
-    def substitute_variables(self, constants=None, variables=None):
-        """Substitute variables of the problem.
+    def substitute_variables(self, constants: SparsePair = None, variables: SparseTriple = None):
+        """Substitutes variables with constants or other variables.
+
+        Args:
+            constants(SparsePair): replace variable by constant
+                i.e., SparsePair.ind (variable) -> SparsePair.val (constant)
+
+            variables(SparseTriple): replace variables by weighted other variable
```

Comment:
If the type hints exist in the signature they can be removed from the docstring.
```suggestion
            constants: replace variable by constant
                i.e., SparsePair.ind (variable) -> SparsePair.val (constant)

            variables: replace variables by weighted other variable
```

Reply:
done